src/server.rs

```diff
@@ -0,0 +1,395 @@
+use std::borrow::Cow;
+use std::io;
+
+use data_encoding::base64;
+use rand::distributions::IndependentSample;
+use rand::distributions::range::Range;
+use rand::os::OsRng;
+use rand::Rng;
+use ring::digest::Digest;
+
+use error::{Error, Kind, Field};
+use utils::find_proofs;
+use ::{NONCE_LENGTH, SHA256_LEN};
+
+/// Responds to client authentication challenges.
+/// The entrypoint for the SCRAM server side implementation
+pub struct ScramServer<P: AuthenticationProvider> {
+    /// The [authentication provider](trait.AuthenticationProvider.html) that will find passwords
+    /// and check authorization
+    provider: P
+
+}
+
+/// Contains information about stored passwords.  In particular, it stores the password that
+/// has been salted and hashed, the salt that was used, and the number of iterations of the hashing
+/// algorithm
+pub struct PasswordInfo {
+    hashed_password: Vec<u8>,
+    salt: Vec<u8>,
```

Comment:
I get your arguments and agree that the ownership of the password (query result) shouldn't be linked to the database connection. I think that the API can be refined later, when the library gets some usage.